### DISCUSSION TITLE
How to dynamically change body_mass

### DISCUSSION THREAD

**Discussion by XinyuWang2** [google-deepmind/mujoco#453] [Asking for Help]
Hi, I need to keep changing some objects' mass in the run time. 
I find the following variable may be relevant:
  physics.model.body_mass
  physics.model.body_inertia

But when I changed it, I find the behavior of the object doesn't change at all (in some way). 
Here is a minimal code that can reproduce the problem. 
In the model(xml), one box is pretty heavy, and another box is trying to push it. 
Even after I change the body mass to 0, it is still unable to be pushed. 
But in some other way, the mass is actually changed! Because if you set the initial pos of box1 to be pos="0 0 0", it will keep flying upward. 

```
import dm_control
import dm_control.mujoco
from dm_control import mjcf
import dm_env
import numpy as np

MODEL = '''
<mujoco model="arena">
    <visual>
        <headlight ambient="0.3 0.3 0.3" diffuse="0.6 0.6 0.6" specular="0 0 0"/>
        <map znear="0.01"/>
    </visual>

    <worldbody>
        <geom name="floor" friction="1.0" type="plane" size=".7 .7 .1" rgba="0.7 0.7 0.7 1"/>
        <body name="box1" pos="0 0 0.1">
            <freejoint/>
            <!-- <geom type="box" size="0.1 0.1 0.1" rgba="1 0 0 0.5" mass="0.01"/> -->
            <geom type="box" size="0.1 0.1 0.1" rgba="1 0 0 0.5" mass="100000"/>
        </body>

        <body name="box2" pos="0 -0.4 0.1">
            <joint name="y" type="slide" axis="0 1 0"/>
            <joint name="z" type="slide" axis="0 0 1"/>
            <geom type="box" size="0.1 0.1 0.1" rgba="0 1 0 0.5" mass="0.1"/>
        </body>

    </worldbody>

    <actuator>
        <motor name="y" joint="y" ctrllimited="true" ctrlrange="-10 10"/>
    </actuator>

</mujoco>
'''


class Env(dm_env.Environment):
    def __init__(self) -> None:
        model = mjcf.from_xml_string(MODEL)
        self.physics = mjcf.Physics.from_mjcf_model(model)

    def step(self, action) -> dm_env.TimeStep:
        self.physics.data.ctrl[:] = action[:]
        self.physics.model.body_mass[1] = 0.0
        # self.physics.model.body_inertia[1] = 0.0
        # print(self.physics.model.body_mass)

        self.physics.step()
        return dm_env.TimeStep(step_type=None,
                                reward=None,
                                discount=None,
                                observation=None)

    def reset(self) -> dm_env.TimeStep:
        return dm_env.TimeStep(step_type=dm_env.StepType.FIRST,
                               reward=None,
                               discount=None,
                               observation=None)
    def action_spec(self):
        action_spec = dm_control.mujoco.action_spec(self.physics)
        return dm_env.specs.BoundedArray(shape=action_spec.shape,
                                        dtype=action_spec.dtype,
                                        minimum=action_spec.minimum,
                                        maximum=action_spec.maximum)
    def observation_spec(self):
        pass

class Policy:

    def __init__(self, env):
        self.action_spec = env.action_spec()

    def __call__(self, time_step):
        action = np.ones(self.action_spec.shape)
        return action

env = Env()
policy = Policy(env)
from dm_control import viewer
viewer.launch(env, policy=policy)
```

**Comment by yuvaltassa** [comment]
Inertia is a [3-vector](https://github.com/deepmind/mujoco/blob/main/include/mujoco/mjmodel.h#L624).
You need to scale the mass and inertia in the same way to get correct physics, so use scaling, not setting:
```
model.body_mass[i] *= c
model.body_inertia[i, :] *= c
```

**Reply by wookayin** [reply]
That is a good point to know! But for this specific example `body_mass` and `body_inertia` are both set to zero, which would have the same effect of scailing.

Out of curiosity, I tried to *scale* the mass and inertia by the same factor, but the red object doesn't move (whose initial mass was like 
 ~100000 and set to a near-zero on runtime).
```
        self.physics.model.body_mass[1] *= 0.00000001
        self.physics.model.body_inertia[1, :] *= 0.00000001
```

Looking at other body values, it looks like there are other simulation states as a consequence of body mass, such as `subtreemass` or `invweight`, etc. In order to correctly change the mass of an object on the fly, do someone need to take care of all these model parameters as well?

**Reply by yuvaltassa** [reply]
Mmm good point. Yes you would need to change those too. Sorry for misleading reply.

**Reply by wookayin** [reply]
It's quite difficult to understand how those subtreemass, invweight, etc. are computed and to which value one should change. They seem to be intertwined values depending on other objects. Can you guide me a pointer to relevant references if any?

For example, with the initial mass of 100000 (for the body 1), subtree/invweight0 has the following initial value (e.g., subtreemass has a high value for the body 0):
```
body_mass [0 100000 0.1]
body_subtreemass [100000 100000 0.1]
body_inertia [[0 0 0] [6.67e+2 6.67e+2 6.67e+2] [6.67e-4 6.67e-4 6.67e-4]
body_invweight0 [[0 0] [0.00001 0.0015] [6.66667 0]]
```

**Reply by fukangl** [reply]
@yuvaltassa When I scale the body mass and inertia using:

`model.body_mass[i] *= c`
`model.body_inertia[i, :] *= c`

There are errors such as 
`WARNING:absl:Inertia matrix is too close to singular at DOF 47. Check model. Time = 0.0000.`
and
`WARNING:absl:Nan, Inf or huge value in QACC at DOF 45. The simulation is unstable. Time = 470.7715.`

The scaling factor is randomly chosen between [0.5,1.5], and even when I use a smaller range [0.8,1.2], this issue still occurs. Do you have any suggestions on how to resolve this problem?

**Reply by yuvaltassa** [reply]
I recommend changing the values in the [mjSpec](https://mujoco.readthedocs.io/en/latest/programming/modeledit.html) and recompiling.

**Reply by fukangl** [reply]
@yuvaltassa Thanks a lot! I use the following code to compile the `mjSpec`. However, no matter what value I use for `self.spec.settotalmass`, the performance of a pre-trained policy (without mass scaling) behaves similarly. I was wondering if this is the correct way to scale the mass and inertia. Any suggestions are appreciated!

```
self.spec = mujoco.MjSpec()
self.spec.from_file(path_to_xml_out)
self.model = self.spec.compile()
self.data = mujoco.MjData(self.model)
self.spec.settotalmass = 50
self.model, self.data = self.spec.recompile(self.model, self.data)
```

**Reply by fukangl** [reply]
During the test, when I use the following code, the policy behaves reasonably. For example, the policy performs best when `c=1` and worse when `c` is much smaller than `1`

```
model.body_mass[i] *= c
model.body_inertia[i, :] *= c
```

**Reply by fukangl** [reply]
It seems this problem has been solved. Just recompile the XML for each training episode, although it requires more memory resources
